framework/src/kernels/ADKernel.C

```diff
@@ -92,8 +92,8 @@ ADKernelTempl<T, compute_stage>::computeResidual()
   prepareVectorTag(_assembly, _var.number());
 
   precalculateResidual();
-  for (_i = 0; _i < _test.size(); _i++)
-    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    for (_i = 0; _i < _test.size(); _i++)
```

Comment:
I don't really see how inverting this loop is a win.  You have timing that shows this?

Reply:
To me lots of operations are only qp dependent, thus move qp loop out can reduce evaluations. Are there any considerations initially to have the original loop?

Reply:
@YaqiWang trading the order of these loops doesn’t change the number of times the function is called.

To do that you will want to use KernelValue or KernelGrad or use “precomputeWhatever()”.

Changing the order of the loops only changes the memory access order.

Doing `_i` first has generally made more sense because things like `test[i][qp]` are then linear through memory.

That’s why I asked if @rwcarlsen has some specific testing that shows that this is faster (it’s possible!  I would just like to have evidence before we change things).

Reply:
Sorry, I was not being clear. I mean that we can provide a virtual function, something like `initQpResidual`, which is in the qp loop. Users can implement it to put qp dependent operations.

Reply:
I'm fairly certain that he has timing to show that this is a win (don't know how big). He's out for the rest of the week so this can wait until next week. We might want to look at the framework though, not just Pronghorn.

Reply:
I have some loop optimization in a pronghorn kernel that (admitedly hacky) does things along the lines of KernelGrad, but to a more extreme extent.  And while I could use KernelGrad, it would make the code even more obfuscated and confusing while not giving quite as big as a performance win.  The point of switching the order here is to make it consistent with the looping nesting order for the corresponding computeOffDiagJacobian function with `qp` being the outer one fore more precompute goodness.

Reply:
1.  I would really like to see timing on this then.  Back in the day we did do timing that showed that the way it currently is is faster.

2.  Let's not do "hacky" things.  If you're storing off values at qps and reusing them - then you should be using `precomputeQpResidual()`.  If you're doing something else then let's talk about it.

Reply:
I'm sorry that should be `precalculateResidual()`

In summary: we definitely don't wan things inside `computeQpResidual()` statements that depend on the ordering of the loops calling that function...

Reply:
precalculateResidual runs outside the qp loop.

Reply:
Yes - but its purpose is to allow you to compute stuff to use inside the qp loop - you can do your own loop over qps if you need to.  This function is sufficiently low-level and focused on optimization that that is fine.

It's outside the loop for optimization - sometimes there is something you can compute that will be the same for all qps.  Or, you may be able to compute part of it for all qps and then compute something for each qp by doing a qp loop.

Also: we wouldn't want another virtual function inside the qp loop that 99% of Kernels wouldn't use.

Reply:
Sure I could do my own qp loop in `precalculateResidual`.  But doesn't it seem odd to have the loop nesting order for `computeADOffDiagJacobian` and `computeResidual` inconsistent?

Reply:
@rwcarlsen and I talked about this.  I'm fine with changing the loop ordering - I would still encourage him not to write code that depends on that ordering though!